hcl/hcl.go

```diff
@@ -1002,6 +1023,115 @@ func parseStack(evalctx *eval.Context, stack *Stack, stackblock *ast.Block) erro
 	return errs.AsError()
 }
 
+func checkHasSubBlocks(block *ast.Block, blocks ...string) error {
+	errs := errors.L()
+
+	for _, got := range block.Attributes.SortedList() {
+		errs.Append(errors.E(got.NameRange,
+			"unrecognized attribute %s.%s", block.Type, got.Name,
+		))
+	}
+
+	for i, label := range block.Labels {
+		errs.Append(errors.E(ErrTerramateSchema,
+			"block %s has unexpected label %s",
+			block.Type,
+			block.LabelRanges[i],
+			label))
+	}
+
+	found := false
+checkBlocks:
+	for _, got := range block.Blocks {
+		for _, want := range blocks {
+			if want == got.Type {
+				if found {
+					errs.Append(errors.E(ErrTerramateSchema,
+						got.DefRange(),
+						"duplicated block %s",
+						got.Type),
+					)
+					continue checkBlocks
+				}
+				found = true
+				continue checkBlocks
+			}
+		}
+
+		errs.Append(errors.E(ErrTerramateSchema,
+			got.DefRange(),
+			"unexpected block %s inside %s",
+			got.Type,
+			block.Type),
+		)
+	}
+
+	return errs.AsError()
+}
+
+func parseVendorConfig(cfg *VendorConfig, vendor *ast.Block) error {
+	logger := log.With().
+		Str("action", "hcl.parseVendorConfig()").
+		Logger()
+
+	if err := checkHasSubBlocks(vendor, "manifest"); err != nil {
+		return err
+	}
+
+	if len(vendor.Blocks) == 0 {
+		return nil
+	}
+
+	manifestBlock := vendor.Blocks[0]
+
+	if err := checkHasSubBlocks(manifestBlock, "default"); err != nil {
+		return err
+	}
+
+	logger.Trace().Msg("parsing vendor.manifest.default block")
+
+	cfg.Manifest = &ManifestConfig{}
+
+	if len(manifestBlock.Blocks) == 0 {
+		return nil
+	}
+
+	defaultBlock := manifestBlock.Blocks[0]
+
+	errs := errors.L()
+
+	for _, block := range defaultBlock.Blocks {
```

Comment:
can't you call `checkHasSubBlocks()` with an empty set of wanted blocks?

Reply:
the function does way too much now, it would also fail because of attributes. I will split them and then it could be used. I started with a schema validation function but it started to be way too much for this case, so I kept changing it to remove some logic and in the end it is an aberration that is not full schema check but also way too complicated :laughing: . Splitting in smaller funcs should make things much more clear.